_posts/2019-09-27-Combining-MongoDB-and-AWS-Lambda.md

```diff
@@ -0,0 +1,272 @@
+
+# Performance: MongoDB vs DynamoDB
+Can MongoDB be an alternative for DynamoDB?  
+In the serverless AWS world DynamoDB is often the database of choice.  
+DynamoDB is a key-value store which is suitable for usecases of storing smaller documents for shorter periods. 
+There certainly are a lot of use cases where you want more functionality then what Dynamo offers you.  
+At that moment you start looking at MongoDB for a solution.
+
+Below I did a performance comparison between DynamoDB and AWS Lambda.
+
+## Test Scenario
+
+In the application below people can creat sessions that they want to give on a conference for software developers.  
+The request to create a sessions is accepted and processed asynchronously.
+The request arrives on a `SNS` topic which fans out towards two `SQS` queues.  
+Both of these queues trigger a Lambda Function:
+* one Lambda Function saves in a DynamoDB table
+* one Lambda Function saves in a MongoDB Atlas collection
+
+I will hit the application with a bunch of sessions that have to be persisted.
+During one minute I will increase the load of incoming sessions from 1/second to 10/second.  
+In total 350 sessions will be fired towards the application during this one minute.
+
+<div style="text-align: center;" >
+  <img src="/img/2019-09-27-Combining-MongoDB-Atlas-and-AWS-Lambda/a-architecture-test.png" width="100%" height="100%">
+</div>
+
+## Performance Results
+I used the AWS services `XRAY` and `CloudWatch` to get insight in the performance.
+
+### Response Distribution
+Xray allows you to get a high level overview of the performance of your functions.  
+It visualizes how fast your function responded over all. 
+Below on the left you see the MongoDB response distribution and on the right the DynamoDB response distribution.
+
+<div style="text-align: center;" >
+  <img src="/img/2019-09-27-Combining-MongoDB-Atlas-and-AWS-Lambda/f-response-distribution-both.png" width="100%" height="100%">
+</div>
+
+This tells us that most of the requests where treated blazingly fast.  
+
+So let's dig deeper.
+
+### Cold Start
+During the cold start the connection with the database is made.  
+When your Lambda container is re-uses, the connection will also be re-used.
+That is, if you implemented it correctly.
+
+Below you see an example of a cold start for both the DynamoDB and MongoDB lambda.  
+
+**They both perform equally concerning the cold start.**
+
+
+<div style="text-align: center;" >
+  <img src="/img/2019-09-27-Combining-MongoDB-Atlas-and-AWS-Lambda/e-dynomo-cold-start.png" width="100%" height="100%">
+</div>
+
+<div style="text-align: center;" >
+  <img src="/img/2019-09-27-Combining-MongoDB-Atlas-and-AWS-Lambda/d-mongodb-cold-start.png" width="100%" height="100%">
+</div>
+
+
+### Performing under load
+As you can see from the images above I created two subsegments representing the work that the MongoDBClient and DynamoDBClient have to do to persist the data towards the database.
+* Sessions.saveSessionDynamoDB
```

Comment:
No image for this graph

Reply:
The image underneath shows the both of them.
This is where you see that the save itself takes less time for MongoDB (at lambda startup). But  the cold start times are quite equal cause setting up the connection with MongoDB takes more time than connecting with DynamoDB